tokio/src/sync/rwlock.rs

```diff
@@ -0,0 +1,244 @@
+
+impl<T> RwLock<T> {
+    /// Creates a new instance of an `RwLock<T>` which is unlocked.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::RwLock;
+    ///
+    /// let lock = RwLock::new(5);
+    /// ```
+    pub fn new(value: T) -> RwLock<T> {
+        RwLock {
+            c: UnsafeCell::new(value),
+            s: Semaphore::new(MAX_READS),
+            w: Mutex::new(()),
+        }
+    }
+
+    /// Locks this rwlock with shared read access, blocking the current task
+    /// until it can be acquired.
+    ///
+    /// The calling task will be blocked until there are no more writers which
+    /// hold the lock. There may be other readers currently inside the lock when
+    /// this method returns.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::sync::Arc;
+    /// use tokio::sync::RwLock;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let lock = Arc::new(RwLock::new(1));
+    ///     let c_lock = lock.clone();
+    ///
+    ///     let n = lock.read().await;
+    ///     assert_eq!(*n, 1);
+    ///
+    ///     tokio::spawn(async move {
+    ///         let r = c_lock.read().await;
+    ///         assert_eq!(*r, 1);
+    ///     })
+    ///}
+    /// ```
+    pub async fn read(&self) -> RwLockReadGuard<'_, T> {
+        let mut permit = ReleasingPermit(Permit::new(), self);
+        poll_fn(|cx| permit.0.poll_acquire(cx, &self.s))
+            .await
+            .unwrap_or_else(|_| {
+                // The semaphore was closed. but, we never explicitly close it, and we have a
+                // handle to it through the Arc, which means that this can never happen.
+                unreachable!()
+            });
+        RwLockReadGuard {
+            lock: self,
+            permit: std::mem::replace(&mut permit.0, Permit::new()),
+        }
+    }
+
+    /// Locks this rwlock with exclusive write access, blocking the current
+    /// task until it can be acquired.
+    ///
+    /// This function will not return while other writers or other readers
+    /// currently have access to the lock.
+    ///
+    /// Returns an RAII guard which will drop the write access of this rwlock
+    /// when dropped.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::RwLock;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///   let lock = RwLock::new(1);
+    ///
+    ///   let mut n = lock.write().await;
+    ///   *n = 2;
+    ///}
+    /// ```
+    pub async fn write(&self) -> RwLockWriteGuard<'_, T> {
+        let _lock = self.w.lock().await;
+
+        let mut permits = vec![];
+        for _ in 0..MAX_READS {
+            let mut permit = ReleasingPermit(Permit::new(), self);
+            poll_fn(|cx| permit.0.poll_acquire(cx, &self.s))
+                .await
+                .unwrap();
+            permits.push(permit);
+        }
+        RwLockWriteGuard {
+            lock: self,
+            permits: permits.split_off(0),
```

Comment:
Same here. Couldn't we just move the whole `ReleasingPermit` into the `RwLockWriteGuard`?

Reply:
followed your example on Discord, and then forgot to remove it after, thanks